packages/did/src/DidBatcher/FullDidBuilder.ts

```diff
@@ -0,0 +1,193 @@
+ *
+ * Concrete classes must subclass this one in order to provide the desired functionality, as is the case with [[FullDidCreationBuilder]] and [[FullDidUpdateBuilder]] handling DID creation and DID updates respectively.
+ */
+export abstract class FullDidBuilder {
+  protected apiObject: ApiPromise
+
+  // New key agreement keys to set as {id -> details}
+  protected newKeyAgreementKeys: Map<
+    DidEncryptionKey['id'],
+    NewDidEncryptionKey
+  > = new Map()
+
+  // New service endpoints to set as {id -> details}
+  protected newServiceEndpoints: Map<
+    DidServiceEndpoint['id'],
+    Omit<DidServiceEndpoint, 'id'>
+  > = new Map()
+
+  // Assertion key action, either ignore, update, or delete. Ignore by default.
+  protected newAssertionKey: VerificationKeyAction = { action: 'ignore' }
+  // Delegation key action, either ignore, update, or delete. Ignore by default.
+  protected newDelegationKey: VerificationKeyAction = { action: 'ignore' }
+
+  protected consumed = false
+
+  public constructor(api: ApiPromise) {
+    this.apiObject = api
+  }
+
+  /**
+   * Mark a new encryption key to be added to the next DID operation.
+   *
+   * The operation will fail in the following cases:
+   *   - The builder has already been consumed
+   *   - There was already a key with the same ID marked for addition.
+   *
+   * @param key The new [[NewDidEncryptionKey]] to add to the DID.
+   * @returns The builder with the provided operation saved internally.
+   */
+  public addEncryptionKey(key: NewDidEncryptionKey): this {
+    if (this.consumed) {
+      throw SDKErrors.ERROR_DID_BUILDER_ERROR(
+        'DID builder has already been consumed.'
+      )
+    }
+    const newKeyId = deriveChainKeyId(this.apiObject, key)
+    // 1. Check if a key with the same ID has already been added.
+    if (this.newKeyAgreementKeys.has(newKeyId)) {
+      throw SDKErrors.ERROR_DID_BUILDER_ERROR(
+        `Key agreement key with ID ${newKeyId} has already been marked for addition. Failing since this may lead to unexpected behaviour.`
+      )
+    }
+    // Otherwise we can safely mark the key for addition.
+    this.newKeyAgreementKeys.set(newKeyId, {
+      publicKey: key.publicKey,
+      type: key.type,
+    })
+    return this
+  }
+
+  /**
+   * Mark a new attestation key to replace the old one, if present, in the next DID operation.
+   *
+   * The operation will fail in the following cases:
+   *   - The builder has already been consumed
+   *   - There was already a new attestation key marked for addition.
+   *
+   * @param key The new [[NewDidVerificationKey]] to add to the DID.
+   * @returns The builder with the provided operation saved internally.
+   */
+  public setAttestationKey(key: NewDidVerificationKey): this {
+    if (this.consumed) {
+      throw SDKErrors.ERROR_DID_BUILDER_ERROR(
+        'DID builder has already been consumed.'
+      )
+    }
+    // 1. Check if another attestation key was already marked for addition.
```

Comment:
Q: If there is a `1.`, where is `2.`? Seeing this multiple times

Reply:
https://github.com/KILTprotocol/sdk-js/pull/473/commits/0419a14aff5a9976e94ca94e598af90f3e2bfb56